hugo2confluence/main.go

```diff
@@ -0,0 +1,330 @@
+			if err != nil {
+				return nil, err
+			}
+			p.children = children
+			pages = append(pages, p)
+			continue
+		}
+
+		// Skip any non-markdown files and _index.md since we handle it differently
+		if entry.Name() == indexFileName || filepath.Ext(fullPath) != ".md" {
+			continue
+		}
+
+		// Add the markdown file as a child with no children
+		pages = append(pages, page{
+			filePath: fullPath,
+		})
+	}
+
+	return pages, nil
+}
+
+func syncPage(api *confluence.API, p page) error {
+	log.Printf("Syncing %s", p.filePath)
+
+	// Generate the content
+	content, err := contentFromFile(p.parentID, p.filePath)
+	if err != nil {
+		return err
+	}
+
+	// Check if the page already exists to determine correct version
+	var existing *confluence.Content
+	inc := 1
+	modifiedTitle := ""
+	for {
+		// If there was a page found with the same name, but the ancestor did not match,
+		// append a number to the end since Confluence pages must have unique names
+		// underneath a root ancestor
+		t := content.Title
+		if inc > 1 {
+			modifiedTitle = fmt.Sprintf("%s (%d)", content.Title, inc)
+			t = modifiedTitle
+		}
+
+		resp, err := api.GetContent(confluence.ContentQuery{
+			Expand:   []string{"version,ancestors"},
+			SpaceKey: env.ConfluenceSpace,
+			Title:    t,
+		})
+		if err != nil {
+			return err
+		}
+		if len(resp.Results) == 0 {
+			break
+		}
+		for _, result := range resp.Results {
+			if result.Ancestors[len(result.Ancestors)-1].ID == p.parentID {
+				existing = &result
+				log.Printf("Found existing page \"%s\" (id:%s)", existing.Title, existing.ID)
+				break
+			}
+		}
+		if existing != nil {
+			break
+		}
+
+		inc++
+	}
+
+	if modifiedTitle != "" {
+		content.Title = modifiedTitle
+	}
+
+	var written *confluence.Content
+	var writeErr error
+
+	if existing != nil {
+		content.ID = existing.ID
+		content.Version.Number = existing.Version.Number + 1
+		log.Printf("Updating page \"%s\" (parent:%s id:%s)", content.Title, p.parentID, content.ID)
+		written, writeErr = api.UpdateContent(content)
+	} else {
+		log.Printf("Creating page \"%s\" (parent:%s)", content.Title, p.parentID)
+		written, writeErr = api.CreateContent(content)
+	}
+
+	if writeErr != nil {
+		return writeErr
+	}
+
+	log.Printf("Wrote page \"%s\" (parent:%s id:%s version:%d url:%s)",
+		written.Title, p.parentID, written.ID, written.Version.Number, written.Links.WebUI)
+
+	// Now recursively write the children
+	// TODO: sync children pages asynchronously?
+	for _, child := range p.children {
+		child.parentID = written.ID
+		if err := syncPage(api, child); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func contentFromFile(parentID string, path string) (*confluence.Content, error) {
+	b, err := os.ReadFile(path)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+	}
+	s := string(b)
+	title := extractTitle(s, path)
+	s = stripMeta(s)
+
+	html := extraHTML(path)
+	if s != "" {
+		converted, err := remoteConvertMarkdown(s)
```

Comment:
Eventually we could have hugo handle the HTML conversion so that it can pull in some extra features

Reply:
Good call. As I was testing, this seemed most reliable way to convert Markdown to an HTML that works well inside Confluence. I'm not sure if it is some HTML variant or pure HTML (haven't done enough research/testing)